cmd/gpu-kubelet-plugin/health.go

```diff
@@ -0,0 +1,143 @@
+	grpc_health_v1.UnimplementedHealthServer
+
+	server *grpc.Server
+	wg     sync.WaitGroup
+
+	regClient registerapi.RegistrationClient
+	draClient drapb.DRAPluginClient
+}
+
+func startHealthcheck(ctx context.Context, config *Config) (*healthcheck, error) {
+	port := config.flags.healthcheckPort
+	if port < 0 {
+		return nil, nil
+	}
+
+	addr := net.JoinHostPort("", strconv.Itoa(port))
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen for healthcheck service at %s: %w", addr, err)
+	}
+
+	regSockPath := (&url.URL{
+		Scheme: "unix",
+		// TODO: this needs to adapt when seamless upgrades
+		// are enabled and the filename includes a uid.
+		Path: path.Join(config.flags.kubeletRegistrarDirectoryPath, DriverName+"-reg.sock"),
+	}).String()
+	klog.V(6).Infof("connecting to registration socket path=%s", regSockPath)
+	regConn, err := grpc.NewClient(
+		regSockPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("connect to registration socket: %w", err)
+	}
+
+	draSockPath := (&url.URL{
+		Scheme: "unix",
+		Path:   path.Join(config.DriverPluginPath(), "dra.sock"),
+	}).String()
+	klog.V(6).Infof("connecting to DRA socket path=%s", draSockPath)
+	draConn, err := grpc.NewClient(
+		draSockPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("connect to DRA socket: %w", err)
+	}
+
+	server := grpc.NewServer()
+	healthcheck := &healthcheck{
+		server:    server,
+		regClient: registerapi.NewRegistrationClient(regConn),
+		draClient: drapb.NewDRAPluginClient(draConn),
+	}
+	grpc_health_v1.RegisterHealthServer(server, healthcheck)
+
+	healthcheck.wg.Add(1)
+	go func() {
+		defer healthcheck.wg.Done()
+		klog.Infof("starting healthcheck service at %s", lis.Addr().String())
+		if err := server.Serve(lis); err != nil {
+			klog.Errorf("failed to serve healthcheck service on %s: %v", addr, err)
+		}
+	}()
+
+	return healthcheck, nil
+}
+
+func (h *healthcheck) Stop() {
+	klog.Info("Stopping healthcheck service")
+	if h.server != nil {
+		h.server.GracefulStop()
+	}
+	h.wg.Wait()
+}
+
+// Check implements [grpc_health_v1.HealthServer].
+func (h *healthcheck) Check(ctx context.Context, req *grpc_health_v1.HealthCheckRequest) (*grpc_health_v1.HealthCheckResponse, error) {
+	knownServices := map[string]struct{}{"": {}, "liveness": {}}
+	if _, known := knownServices[req.GetService()]; !known {
+		return nil, status.Error(codes.NotFound, "unknown service")
+	}
+
+	status := &grpc_health_v1.HealthCheckResponse{
+		Status: grpc_health_v1.HealthCheckResponse_NOT_SERVING,
+	}
+
+	info, err := h.regClient.GetInfo(ctx, &registerapi.InfoRequest{})
+	if err != nil {
+		klog.ErrorS(err, "failed to call GetInfo")
+		return status, nil
+	}
+	klog.V(6).Infof("Successfully invoked GetInfo: %v", info)
+
+	_, err = h.draClient.NodePrepareResources(ctx, &drapb.NodePrepareResourcesRequest{})
```

Comment:
What is this after all doing? And what is this signaling when it succeeds?
CC @klueska